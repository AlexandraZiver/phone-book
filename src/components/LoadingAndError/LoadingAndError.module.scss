@import "../../styles/mixins/header.scss";

.ContainerError {
  font-size: 18px;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  
}

.LoadingContainer {
  position: relative;
  height: 100vh;
}

.ContainerErrorBig {
  @include header();
  font-size: 18px;
  text-align: center;
}

.LoadingContainerBig .ui.dimmer {
  left: auto !important;
}

