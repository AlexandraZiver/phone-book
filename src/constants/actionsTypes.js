const GET_CLIENTS = "GET_ALL_CLIENTS";

export { GET_CLIENTS };
