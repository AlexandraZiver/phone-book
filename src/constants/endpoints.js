const BASE_URL = "http://localhost:8000/clients";
export { BASE_URL };
