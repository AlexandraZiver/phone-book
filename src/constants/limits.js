const LIMIT_VISIBLE_ITEMS = 13;
export { LIMIT_VISIBLE_ITEMS };
