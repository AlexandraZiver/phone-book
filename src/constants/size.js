const Size = {
  BIG: "big",
  SMALL: "small",
};

export { Size };
