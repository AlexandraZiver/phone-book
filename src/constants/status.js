const Status = {
  INITIAL: null,
  PENDING: "pending",
  REJECTED: "rejected",
};
export default Status;
