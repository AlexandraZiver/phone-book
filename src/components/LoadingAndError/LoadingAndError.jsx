import cx from "classnames";
import PropTypes from "prop-types";
import React from "react";
import { Loader, Dimmer } from "semantic-ui-react";

import { Size } from "../../constants/size";
import styles from "./LoadingAndError.module.scss";

const LoadingAndError = ({ children, status, error, size, text }) => {
  const loading = (text) => {
    if (text) {
      return <>Loading</>;
    }

    return null;
  };

  const containerClassNames = cx({
    [styles.ContainerErrorBig]: size == Size.Big && status == "error",
    [styles.ContainerError]: status == "error",
    [styles.LoadingContainerBig]: status == "loading" && size == Size.BIG,
    [styles.LoadingContainer]: status == "loading" && size == Size.SMALL,
  });

  if (status === "loading") {
    return (
      <div className={containerClassNames}>
        <Dimmer active>
          <Loader>{loading(text)}</Loader>
        </Dimmer>
      </div>
    );
  }
  if (status === "error") {
    return <div className={containerClassNames}>{error}</div>;
  }

  return <>{children}</>;
};

LoadingAndError.defaultProps = {
  text: false,
};

LoadingAndError.propTypes = {
  children: PropTypes.node,
  status: PropTypes.string,
  error: PropTypes.string,
  dependence: PropTypes.bool,
  size: PropTypes.string,
  text: PropTypes.bool,
};

export default LoadingAndError;
