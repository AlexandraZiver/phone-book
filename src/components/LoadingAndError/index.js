import LoadingAndError from "./LoadingAndError";

export default LoadingAndError;
