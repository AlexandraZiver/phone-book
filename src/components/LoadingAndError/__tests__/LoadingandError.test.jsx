import { render } from "@testing-library/react";
import LoadingAndError from "../LoadingAndError";
import Status from "../../../constants/status";
import { Size } from "../../../constants/size";

jest.mock("semantic-ui-react", () => ({
  Loader: jest.fn().mockImplementation(() => <div>Loader</div>),
  Dimmer: jest.fn().mockImplementation(() => <div>Loading...</div>),
}));

describe("LoadingAndError testing", () => {
  test(" renders LoadingAndError and  children", () => {
    const { getByTestId } = render(
      <LoadingAndError>
        <div data-testid="childComponent">Child component</div>
      </LoadingAndError>,
    );

    expect(getByTestId("childComponent")).toBeInTheDocument();
    expect(getByTestId("childComponent")).toMatchSnapshot();
  });

  test("renders LoadingAndError when status is PENDING and size BIG ", () => {
    const { container } = render(
      <LoadingAndError status={Status.PENDING} size={Size.Big} includeText={true} />,
    );

    const loadingElement = container.querySelector(".LoadingContainer", ".LoadingContainerBig");

    expect(loadingElement).toBeInTheDocument();
    expect(loadingElement).toHaveTextContent("Loading...");
    expect(loadingElement).toMatchSnapshot();
  });

  test("renders LoadingAndError when status is REJECTED and size BIG ", () => {
    const { container } = render(
      <LoadingAndError size={Size.BIG} status={Status.REJECTED} error="error" />,
    );

    const errorElement = container.querySelector(".ContainerErrorBig", ".ContainerError");

    expect(errorElement).toBeInTheDocument();
    expect(errorElement).toHaveTextContent("error");
    expect(errorElement).toMatchSnapshot();
  });

  test(" renders LoadingAndError when status REJECTED and size SMALL", () => {
    const { container } = render(<LoadingAndError status={Status.REJECTED} />);

    expect(container.querySelector(".ContainerError")).toBeInTheDocument();
    expect(container.querySelector(".ContainerError")).toMatchSnapshot();
  });

  test(" renders LoadingAndError when status PENDING and size SMALL", () => {
    const { container } = render(<LoadingAndError status={Status.PENDING} size={Size.SMALL} />);

    expect(container.querySelector(".LoadingContainer")).toBeInTheDocument();
    expect(container.querySelector(".LoadingContainer")).toMatchSnapshot();
  });

  test(" renders LoadingAndError when incorrect props", () => {
    const { getByTestId } = render(
      <LoadingAndError size="Large" error="error" includeText={true}>
        <div data-testid="childComponent">Child component</div>
      </LoadingAndError>,
    );

    expect(getByTestId("childComponent")).toBeInTheDocument();
    expect(getByTestId("childComponent")).not.toHaveClass();
    expect(getByTestId("childComponent")).toMatchSnapshot();
  });
});
