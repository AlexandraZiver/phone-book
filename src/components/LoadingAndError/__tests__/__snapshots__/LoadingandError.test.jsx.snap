// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingAndError testing  renders LoadingAndError and  children 1`] = `
<div
  data-testid="childComponent"
>
  Child component
</div>
`;

exports[`LoadingAndError testing  renders LoadingAndError when incorrect props 1`] = `
<div
  data-testid="childComponent"
>
  Child component
</div>
`;

exports[`LoadingAndError testing  renders LoadingAndError when status PENDING and size SMALL 1`] = `
<div
  class="LoadingContainer"
>
  <div>
    Loading...
  </div>
</div>
`;

exports[`LoadingAndError testing  renders LoadingAndError when status REJECTED and size SMALL 1`] = `
<div
  class="ContainerError"
/>
`;

exports[`LoadingAndError testing renders LoadingAndError when status is PENDING and size BIG  1`] = `
<div
  class="LoadingContainer"
>
  <div>
    Loading...
  </div>
</div>
`;

exports[`LoadingAndError testing renders LoadingAndError when status is REJECTED and size BIG  1`] = `
<div
  class="ContainerErrorBig ContainerError"
>
  error
</div>
`;
